### PR TITLE
Aravind's code

### PR DESCRIPTION
**Pull request by aravindrsripada**
My code to be reviewed!

### CODE REVIEW

File: src/test/java/StringManipulationTest.java [hunk 1]
```diff
@@ -20,66 +27,93 @@ public void tearDown() {
     }
```
Comment (Callowlock) [inline src/test/java/StringManipulationTest.java:28]: Aravind, you're missing a requirement in your tests to verify that getString() is returning null when the string is null.

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,32 +1,100 @@
+/*
+Aravind Sripada
+CS410 - Sara Farag
+6/13/2023
+This file contains the implementations for the string manipulations class.
+*/
 public class StringManipulation implements StringManipulationInterface {
```
Comment (Callowlock) [inline src/main/java/StringManipulation.java:12]: No issues with the build on TravisCI.

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,32 +1,100 @@
...
+            return 0;
+
+        String[] wordcount = current.trim().split(" "); //gets rid of whitespace from the front and back
+        return wordcount.length;
     }
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+        if(current == null || current.isEmpty())
```
Comment (Callowlock) [inline src/main/java/StringManipulation.java:42]: Great job checking for current == null before trying to call a method on the null string. 
It would error if you had a null string and current.isEmpty() came before current == null in your conditions.

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,32 +1,100 @@
...
+
+        int wordcount = count();
+        if(startWord > wordcount || endWord > wordcount) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        int start = startWord - 1; //the given ints are always 1 ahead since they start counting from 1
+        int end = endWord - 1;
+
+        String[] words = current.trim().split(" "); //removes white space at the beginning and end of the string and splits by whitespace
```
Comment (Callowlock) [inline src/main/java/StringManipulation.java:81]: I don't see any issue with using this method to break up the words in a String, but instead of instantiating another array and filling it, you could use a regex to extract and populate the output array directly.

File: src/main/java/StringManipulation.java [hunk 5]
```diff
@@ -1,32 +1,100 @@
...
 
     @Override
     public void setString(String string) {
+        current = string;
     }
 
     @Override
     public int count() {
-        return 0;
+        if(current == null || current.isEmpty()) //empty or null string have a length of 0
```
Comment (Callowlock) [inline src/main/java/StringManipulation.java:26]: You may want to consider throwing a NullPointerException when count, or any of your other methods are called on a null string instead of returning something.

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -1,32 +1,100 @@
...
+    public String restoreString(int[] indices){
+        if(indices.length != current.length()) {
+            throw new IllegalArgumentException();
+        }
+
+
+
+        String newStr = "";
+        for(int i = 0; i < indices.length; i++) {
+            if(indices[i] < 0 || indices[i] > current.length() - 1) {
```
Comment (Callowlock) [inline src/main/java/StringManipulation.java]: Something minor: 
I think it's a little clearer if you use "indices[i] >= current.length()" instead of "indices[i] > current.length() - 1".

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -20,66 +27,93 @@ public void tearDown() {
...
 
     @Test
+    //this test check for correct output
     public void testCount4() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("hunkady hunkad hunka hunk hun hu h");
+        int length = manipulatedstring.count();
+        assertEquals(7, length);
     }
```
Comment (Callowlock) [inline src/test/java/StringManipulationTest.java:66]: I don't see a test case for strings that look like "!!!!! !!!!". Your count method would return 2 words in this case and a group of exclamation points is not a word.

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -20,66 +27,93 @@ public void tearDown() {
...
 
     @Test
```
Comment (Callowlock) [inline src/test/java/StringManipulationTest.java:67]: You have quite a few of your test methods marked "throws Exception." This is unnecessary. You should remove the throws statement in the method declaration.

File: src/test/java/StringManipulationTest.java [hunk 9]
```diff
@@ -20,66 +27,93 @@ public void tearDown() {
...
         manipulatedstring.setString("I'd b3tt3r put s0me d161ts in this 5tr1n6, right?");
         assertEquals("I'  b tt r  ut s0 e  16 ts in th s  tr n6  r gh ?", manipulatedstring.removeNthCharacter(3, true));
     }
 
     @Test
-    public void testRemoveNthCharacter3() {
-        fail("Not yet implemented");
+    //this test check for correct output by leaving the string with its default value
+    public void testRemoveNthCharacter3() throws Exception {
+        assertEquals(null, manipulatedstring.removeNthCharacter(3, false));
```
Comment (Callowlock) [inline src/test/java/StringManipulationTest.java:84]: Instead of "assertEquals" you should use "assertNull" here.
Another option is (as mentioned in the StringManipulation file) is set your methods to throw a NullPointerException when your string field is null. In this case you can assertThrows(NullPointerException.class, ....).

File: src/test/java/StringManipulationTest.java (hunk 8 above)
Comment (Callowlock) [reply src/test/java/StringManipulationTest.java:67]: image

Comment (Callowlock) [reply src/test/java/StringManipulationTest.java:67]: As you can see from my Intellij, it doesn't recognize the method as throwing an exception.
The test cases will also run just fine without the throws Exception statement.

File: src/test/java/StringManipulationTest.java [hunk 10]
```diff
@@ -20,66 +27,93 @@ public void tearDown() {
...
     }
 
     @Test
-    public void testRemoveNthCharacter7() {
-        fail("Not yet implemented");
+    //this test check for correct output
+    public void testRemoveNthCharacter7() throws Exception {
+        manipulatedstring.setString("CS4420");
+        assertEquals("CS420", manipulatedstring.removeNthCharacter(4, false));
     }
```
Comment (Callowlock) [inline src/test/java/StringManipulationTest.java:120]: One note on the getSubstrings() method:
I notice that you don't have test cases that include punctuation, like "I need some sleep!". The reason I mention this is that I can tell by the way you split strings, the "!" will be included with the string. I'd recommend using a regex to extract your words, or additional processing on the temp array you use to hold strings to remove punctuations.

**Review by Callowlock — CHANGES_REQUESTED** [review]
Aravind,

Very nice start on your code. 
The only branch in your code that was missed with a test case was the getSubstring() return null, which I already mentioned. The other big issue I found was marking your test cases with "throws Exception". These is the only issues I would explicitly say that you should fix before merging. 
Other than that, I would suggest adjusting your getSubstring() method to drop punctuations from words you extract, adjusting your count() method to exclude "words" made of punctuations (!!!! !!!! ----), and maybe changing your methods to throw a NullPointerException when called on a null string. These are dependent on the client's requirements though.

image
